### PR TITLE
[WPB-10783] Prevent MLS-Legalhold interactions

### PR DESCRIPTION
**Pull request by MangoIV**
## Checklist

 - [x] Add a new entry in an appropriate subdirectory of `changelog.d`
 - [x] Read and follow the [PR guidelines](https://docs.wire.com/developer/developer/pr-guidelines.html)

### CODE REVIEW

File: integration/test/Test/LegalHold.hs [hunk 1]
```diff
@@ -906,15 +906,33 @@ testLHDisableBeforeApproval = do
...
+  (charlie, tid, []) <- createTeam OwnDomain 1
+  [charlie1] <- traverse (createMLSClient def) [charlie]
+  void $ createNewGroup charlie1
+  void $ createAddCommit charlie1 [charlie] >>= sendAndConsumeCommitBundle
+
+  legalholdWhitelistTeam tid charlie >>= assertStatus 200
+  withMockServer def lhMockApp \lhDomAndPort _chan -> do
+    postLegalHoldSettings tid charlie (mkLegalHoldSettings lhDomAndPort) >>= assertStatus 201
+    requestLegalHoldDevice tid charlie charlie `bindResponse` do
+      assertLabel 409 "mls-legal-hold-not-allowed"
```
Comment (MangoIV) [inline integration/test/Test/LegalHold.hs:925]: This is nice ~ denying the request seems like the better option.

Comment (MangoIV) [reply integration/test/Test/LegalHold.hs:925]:
```suggestion
    requestLegalHoldDevice tid charlie charlie `bindResponse`
      assertLabel 409 "mls-legal-hold-not-allowed"
```

File: services/galley/src/Galley/API/LegalHold.hs [hunk 2]
```diff
@@ -392,6 +394,12 @@ requestDevice lzusr tid uid = do
     lhs@UserLegalHoldDisabled -> RequestDeviceSuccess <$ provisionLHDevice zusr luid lhs
     UserLegalHoldNoConsent -> throwS @'NoUserLegalHoldConsent
   where
+    disallowIfMLSUser :: Local UserId -> Sem r ()
+    disallowIfMLSUser luid = do
+      void $ iterateConversations luid (toRange (Proxy @500)) $ \convs -> do
+        when (any (\c -> c.convProtocol /= ProtocolProteus) convs) $ do
+          throwS @'MLSLegalholdIncompatible
```
Comment (MangoIV) [inline services/galley/src/Galley/API/LegalHold.hs:401]:
```suggestion
    disallowIfMLSUser luid =
      void $ iterateConversations luid (toRange (Proxy @500)) $ \convs ->
        when (any (\c -> c.convProtocol /= ProtocolProteus) convs) $
          throwS @'MLSLegalholdIncompatible
```

File: services/galley/src/Galley/API/LegalHold.hs [hunk 3]
```diff
@@ -401,6 +409,7 @@ requestDevice lzusr tid uid = do
     -- device at (almost) the same time.
     provisionLHDevice :: UserId -> Local UserId -> UserLegalHoldStatus -> Sem r ()
     provisionLHDevice zusr luid userLHStatus = do
+      disallowIfMLSUser luid
```
Comment (MangoIV) [inline services/galley/src/Galley/API/LegalHold.hs:412]: maybe a small comment would be nice as well, even though I won't insist on it, the code is already quite clear. 
```suggestion
      -- currently, legalhold is not compatible with MLS, hence we disallow provisioning
      -- a legalhold device for any user who partakes in anything but proteus conversations. 
      disallowIfMLSUser luid
```

**Review by MangoIV — COMMENTED** [review]
looks good to me. I cannot approve as this is my PR but consider it approved.

File: integration/test/Test/LegalHold.hs (hunk 1 above)
Comment (akshaymankar) [reply integration/test/Test/LegalHold.hs:925]: Format becomes very ugly without the do.

File: services/galley/src/Galley/API/LegalHold.hs (hunk 3 above)
Comment (akshaymankar) [reply services/galley/src/Galley/API/LegalHold.hs:412]: Took long enough for CI, I think I won't add it. Feel free to create another PR.

File: services/galley/src/Galley/API/LegalHold.hs (hunk 2 above)
Comment (akshaymankar) [reply services/galley/src/Galley/API/LegalHold.hs:401]: Gonna leave these too, CI took too long, I'm tired today.

**Review by akshaymankar — APPROVED** [review]

**Pull request merged**